hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs

```diff
@@ -835,3 +835,178 @@ replaceContesters contesters = \case
+  | CouldNotFindTheCorrectCurrencySymbolInTokens
+  | MultipleHeadTokensOrMoreThan1PTsFound
+  | NothingCommittedButTxOutInOutputDatum
+  | CommittedTxOutButNothingInOutputDatum
+  | MissingDatum
+  | UnexpectedInlineDatum
+  | CouldNotFindDatum
+  | ExpectedCommitDatumTypeGotSomethingElse
+  | ExpectedSingleCommitOutput
+  deriving (Show)
+
+data CommitError
+  = STNotBurnedError
+  | STIsMissingInTheOutput
+  deriving (Show)
+
+data HeadError
+  = InvalidHeadStateTransition
+  | BurntTokenNumberMismatch
+  | ReimbursedOutputsDontMatch
+  | STNotSpent
+  | IncorrectUtxoHash
+  | ChangedParameters
+  | WrongStateInOutputDatum
+  | MissingCommits
+  | HeadValueIsNotPreserved
+  | HasBoundedValidityCheckFailed
+  | InvalidSnapshotSignature
+  | ClosedWithNonInitialHash
+  | IncorrectClosedContestationDeadline
+  | InfiniteUpperBound
+  | InfiniteLowerBound
+  | ContestersNonEmpty
+  | TooOldSnapshot
+  | UpperBoundBeyondContestationDeadline
+  | ContestNoUpperBoundDefined
+  | MustNotPushDeadline
+  | MustPushDeadline
+  | ContesterNotIncluded
+  | WrongNumberOfSigners
+  | SignerAlreadyContested
+  | FannedOutUtxoHashNotEqualToClosedUtxoHash
+  | LowerBoundBeforeContestationDeadline
+  | FanoutNoLowerBoundDefined
+  | CloseNoUpperBoundDefined
+  | ScriptNotSpendingAHeadInput
+  | SignerIsNotAParticipant
+  | NoSigners
+  | TooManySigners
+  | NoOutputDatumError
+  | DatumNotFound
+  | SignatureVerificationFailed
+  | PartySignatureVerificationFailed
+  deriving (Show)
+
+class Show a => ToErrorCode a where
+  toCode :: a -> Text
+
+instance ToErrorCode UtilError where
+  toCode :: UtilError -> Text
+  toCode = \case
+    MintingOrBurningIsForbidden -> "U01"
+
+instance ToErrorCode HeadTokensError where
+  toCode :: HeadTokensError -> Text
+  toCode = \case
+    SeedNotSpent -> "M01"
+    WrongNumberOfTokensMinted -> "M02"
+    MissingST -> "M03"
+    WrongNumberOfInitialOutputs -> "M04"
+    WrongDatum -> "M05"
+    MintingNotAllowed -> "M06"
+    NoPT -> "M07"
+    WrongQuantity -> "M08"
+    HeadDatum -> "M09"
+    NoDatum -> "M10"
+    MultipleHeadOutput -> "M11"
+
+instance ToErrorCode InitialError where
+  toCode :: InitialError -> Text
+  toCode = \case
+    STNotBurned -> "I01"
+    MissingOrInvalidCommitAuthor -> "I02"
+    LockedValueDoesNotMatch -> "I03"
+    MismatchCommittedTxOutInDatum -> "I04"
+    CouldNotFindTheCorrectCurrencySymbolInTokens -> "I05"
+    MultipleHeadTokensOrMoreThan1PTsFound -> "I06"
+    NothingCommittedButTxOutInOutputDatum -> "I07"
+    CommittedTxOutButNothingInOutputDatum -> "I08"
+    MissingDatum -> "I09"
+    UnexpectedInlineDatum -> "I10"
+    CouldNotFindDatum -> "I11"
+    ExpectedCommitDatumTypeGotSomethingElse -> "I12"
+    ExpectedSingleCommitOutput -> "I13"
+
+instance ToErrorCode CommitError where
+  toCode :: CommitError -> Text
+  toCode = \case
+    STNotBurnedError -> "C01"
+    STIsMissingInTheOutput -> "C02"
+
+instance ToErrorCode HeadError where
+  toCode :: HeadError -> Text
```

Comment:
Potentially we could generate these instances using TH

Reply:
Using TH was the idea here, but I do not know how, so we decided to open this PR in a dirty way first.

Do you think we should do it in this PR?

Reply:
From our tactical meeting, we agreed that we will tackle this in a different PR.